### PR TITLE
SerialPort test improvements

### PR DESCRIPTION
**Pull request by willdean**
This is a substantial set of changes which need to bake against the CI a bit before we consider merging.

* More misc cleanup of migrated tests
* Fixes for test bugs introduced by me during the migration
* Addition of hardware buffer probing and associated buffer size selection (addresses #16033)
* Experimental removal of some suppression which should be fixed by the buffer probing 
* Removal of tests which solely test hardware capabilities and don't enhance coverage of SerialPort

The tests I have removed have the following characteristics:

1. They test for the presence of archaic serial port features (5/6 bit data, 1.5 stop bits) which are no longer in general use and are not supported by some of the popular modern USB-Serial devices - so we get test failures there even though there is no fault in SerialPort.
2. They don't improve coverage of SerialPort at all - they merely represent different numbers for it to pass down through to the underlying stack.
3. We have other tests which verify the correct behaviour of SerialPort when the underlying stack rejects values for these parameters that it doesn't like.
4. One of them assumed that a baud rate of 12345 can't be selected, just it doesn't work on the original 16550 hardware (which is a clone of the 8250 in a PC/XT) - there is nothing fundamental which says that a port can't offer that rate, and indeed the USB-Serial solutions often can.

I don't consider these tests to have any value, given that we'll have to be able to skip them for people to be able to run the tests against modern hardware anyway *and they offer no additional coverage*.

Particularly as we move to a more cross-platform approach, SerialPort's test suite is going to have to be more broadminded about what specific features of serial ports might be available - as it stands, the test suite is second-guessing what features the stack offers, based on a very narrow set of assumptions.

With this PR, we now have a complete null-modem test pass using conventional hardware (16550 clone on motherboard as first port,  16c950 on PCI card as second port) - this is a major milestone.   We also have much better pass rate on USB-Serial hardware than we did, though it's not 100% yet.

**System.IO.Ports.Tests  Total: 995, Errors: 0, Failed: 0, Skipped: 0, Time: 1145.502s**

### CODE REVIEW

**Comment by danmosemsft** [comment]
You probably noticed but there are a lot of Serial Port failures right now helping give us a red badge:
https://ci.dot.net/job/dotnet_corefx/job/master/job/outerloop_win7_debug/lastCompletedBuild/testReport/

**Comment by willdean** [comment]
@danmosemsft Yes, I awoke this AM to a #16100 and #16099 and filed #16102 as a nuclear solution - it disables all the hardware-requiring tests on both inner- and outer-loop until we can stabilise them.

This PR is different, because I'm trying to stabilise the tests on it, so I have them enabled.

**Comment by danmosemsft** [comment]
Sounds good, I"m trying to get those to green. When we need to experiment with  how they do in Jenkins, I believe there's a way to do private runs, which I can dig up.

**Comment by danmosemsft** [comment]
Re, private runs I have this note, but didn't try it myself:

> You don’t need to submit a PR to be able to execute something in CI.  You can just go to:
> https://ci.dot.net/job/dotnet_corefx/job/master/job/windows_nt_release_prtest/build?delay=0sec
> (this is the link for the Windows_NT release job, but you can pick a different one from Jenkins if you want).  In the first box, enter the name of your test branch in your repo, in the second box include the link to your repo, clear the third box so that it’s blank, and leave the fourth as is, then hit Build.  You’ll see a new job appear on the bottom left.

@mmitche would this run outer loop or inner loop? (We need outer loop)

**Comment by mmitche** [comment]
@danmosemsft That's inner loop, for outer you'd want: https://ci.dot.net/job/dotnet_corefx/job/master/job/outerloop_windows_nt_release_prtest/build?delay=0sec

**Comment by danmosemsft** [comment]
Ah - right, thanks. @willdean let us know how it goes. Anything you don't plan to fix today, please let's diable with an issue number so we can get the run clean.

**Comment by willdean** [comment]
@danmosemsft Thanks for that about the build - I can play with that in future.

If we get #16102 merged, that will fix all your test problems anyway.

**Comment by willdean** [comment]
@dotnet-bot test Outerloop Windows_NT Debug please

**Comment by willdean** [comment]
@dotnet-bot test Outerloop Windows_NT Debug please

**Comment by danmosemsft** [comment]
@dotnet-bot test outerloop windows_nt debug please (https://github.com/dotnet/corefx/issues/14838)

@willdean something I try to do, which is optional, is include a link to the issue when I rerun CI. By definition, when we rerun manually, there is some unrelated flaky test. By linking to the issue, it shows that htere is one, and also the issue it gets another link on it, showing it's a big problem.

File: src/System.IO.Ports/tests/Legacy/Support/TCSupport.cs [hunk 1]
```diff
@@ -605,5 +622,71 @@ public static void VerifyArray<T>(T[] expectedArray, T[] actualArray, int index,
...
+        /// <summary>
+        /// Wait for the write data to be written into a blocked (by adverse flow control) port
+        /// </summary>
+        public static void WaitForWriteBufferToLoad(SerialPort com, int bufferLength)
+        {
+            Stopwatch sw = Stopwatch.StartNew();
+            while (com.BytesToWrite + HardwareTransmitBufferSize < bufferLength)
+            {
+                System.Threading.Thread.Sleep(50);
+                if (sw.ElapsedMilliseconds > 3000)
```
Comment (JeremyKuhne) [inline src/System.IO.Ports/tests/Legacy/Support/TCSupport.cs]: You should inline these checks into the Assert.  It's OK to Assert.True(false), etc. where it requires work to refactor. The True/False ones allow you to pass the descriptive text.

Comment (willdean) [inline src/System.IO.Ports/tests/Legacy/Support/TCSupport.cs]: Yes, good point - I don't know why I didn't do that there - if you look a few lines further down in that file I've done exactly what you said - I'll tidy that up.

**Comment by willdean** [comment]
@danmosemsft At the moment it feels like I wouldn't be far off putting #* at the end of every CI rerun, but I take your point.   TBH I'm not really CI re-running in this PR because of *unrelated* flaky tests.  I'm doing it because I'm trying to reduce the flakiness of this PR when running on CI.

**Comment by danmosemsft** [comment]
Fair enough. :)

**Comment by willdean** [comment]
@dotnet-bot test outerloop windows_nt debug please (pre-merge checks)

**Comment by willdean** [comment]
@JeremyKuhne I think, subject to any further review, we could merge this.   All the hardware tests are now disabled on the CI anyway, so I think it's very low-risk in terms of breaking the build.

Then maybe we could look at a PR to move these to the proper directories and get them all namespaced properly - but I fell like I'd quite like to do that separately to any actual code changes.

What do you think?

**Comment by JeremyKuhne** [comment]
@willdean I was just waiting for your assertion that you were ready. As long as we're keeping the history intact with filenames and namespaces I don't care when that happens. Thanks again for your work here!

**Pull request merged**